### PR TITLE
feat(genesis bridge): genesis transfers

### PR DESCRIPTION
**Pull request by danwt**
## Description

See

https://github.com/dymensionxyz/research/issues/278 for the master issue.

There are a few outstanding items, that this PR does not try to solve.

This PR:

- do not allow opening any ibc connections starting at the RA side
- hard codes alice/bob keys in test scripts to make the address deterministic
- adds `-y` to register rollapp / register sequencer cmds
- wires in the new hub genesis ibc module (see RDK pr https://github.com/dymensionxyz/dymension-rdk/pull/449)

NOTE: I will update the go.mod to the correct dependencies after the other prs are merged

----

Closes https://github.com/dymensionxyz/rollapp-evm/issues/278

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

## Summary by CodeRabbit

- **New Features**
  - Introduced a message type check to prevent connection openings from one side.

- **Improvements**
  - Updated scripts for setting up IBC paths and registering rollapps, sequencers, and genesis accounts with new flags for automation and better organization.
  - Enhanced genesis file updates for account balances, metadata, and elevated addresses.

- **Permissions**
  - Modified permissions for specific module names.
  - Adjusted `TransferKeeper` initialization for better control.

### CODE REVIEW

File: scripts/trigger_rollapp_genesis_event.sh
Comment (srene) [inline scripts/trigger_rollapp_genesis_event.sh:1]: i think we need to update the readme file to be aligned with the changes. also if you could update the block time params ...

File: go.mod [hunk 1]
```diff
@@ -331,3 +331,8 @@ replace (
 	// use cometbft
 	github.com/tendermint/tendermint => github.com/cometbft/cometbft v0.34.28
 )
+
```
[the remark this answers is not in the document]
Comment (danwt) [reply go.mod:334]: I will after the others prs are merged

File: scripts/trigger_rollapp_genesis_event.sh
Comment (danwt) [reply scripts/trigger_rollapp_genesis_event.sh:1]: will do

**Comment by danwt** [comment]
Thanks @srene I've fixed the readme now

File: README.md [hunk 2]
```diff
@@ -147,17 +145,12 @@ Modify `dymint.toml` in the chain directory (`~/.rollapp_evm/config`)
...
-sed -i '/rollapp_id =/c\rollapp_id = '\"$ROLLAPP_CHAIN_ID\" "${ROLLAPP_HOME_DIR}/config/dymint.toml"
-```
-
-mac:
-
-```shell
-sed -i '' 's/settlement_layer.*/settlement_layer = "dymension"/' ${ROLLAPP_HOME_DIR}/config/dymint.toml
-sed -i '' 's|node_address =.*|node_address = '\"$HUB_RPC_URL\"'|' "${ROLLAPP_HOME_DIR}/config/dymint.toml"
-sed -i '' 's|rollapp_id =.*|rollapp_id = '\"$ROLLAPP_CHAIN_ID\"'|' "${ROLLAPP_HOME_DIR}/config/dymint.toml"
+dasel put -f "${ROLLAPP_HOME_DIR}"/config/dymint.toml "settlement_layer" -v "dymension"
```
Comment (mtsitrin) [inline README.md:148]: not sure we need to introduce additional dependency here
i'd keep the `sed` cmd

**Review by mtsitrin — DISMISSED** [review]

File: scripts/ibc/setup_ibc.sh [hunk 3]
```diff
@@ -95,7 +95,7 @@ fi
 
 echo '--------------------------------- Creating IBC path... --------------------------------'
 
-rly paths new "$ROLLAPP_CHAIN_ID" "$SETTLEMENT_CHAIN_ID" "$RELAYER_PATH" --src-port "$IBC_PORT" --dst-port "$IBC_PORT" --version "$IBC_VERSION"
+rly paths new "$SETTLEMENT_CHAIN_ID" "$ROLLAPP_CHAIN_ID" "$RELAYER_PATH" --src-port "$IBC_PORT" --dst-port "$IBC_PORT" --version "$IBC_VERSION"
```
Comment (danwt) [inline scripts/ibc/setup_ibc.sh:99]: @zale144

File: README.md [hunk 4]
```diff
@@ -147,17 +145,12 @@ Modify `dymint.toml` in the chain directory (`~/.rollapp_evm/config`)
 linux:
```
Comment (zale144) [inline README.md:145]: will this work on mac?

File: scripts/settlement/register_rollapp_to_hub.sh [hunk 5]
```diff
@@ -15,19 +15,13 @@ if [ -z "$HUB_CHAIN_ID" ]; then
...
 set -x
 #Register rollapp
 dymd tx rollapp create-rollapp "$ROLLAPP_CHAIN_ID" "$MAX_SEQUENCERS" '{"Addresses":[]}' \
-  "$DENOM_METADATA_PATH" \
-  --genesis-accounts-path "$GENESIS_ACCOUNTS_PATH" \
   --from "$DEPLOYER" \
   --keyring-backend test \
   --broadcast-mode block \
-  --fees 1dym
+  --fees 1dym \
```
Comment (zale144) [inline scripts/settlement/register_rollapp_to_hub.sh:24]: shouldn't this also have `--transfers-enabled true` ?

File: README.md (hunk 4 above)
Comment (danwt) [reply README.md:145]: sorry yes meant to delete the linux word

File: scripts/settlement/register_rollapp_to_hub.sh (hunk 5 above)
Comment (danwt) [reply scripts/settlement/register_rollapp_to_hub.sh:24]: no
but now it's different anyway after
https://github.com/dymensionxyz/dymension/pull/961
https://github.com/dymensionxyz/dymension-rdk/pull/458

File: app/ante/handlers.go [hunk 6]
```diff
@@ -43,6 +44,7 @@ func cosmosHandler(options HandlerOptions, sigChecker sdk.AnteDecorator) sdk.Ant
 		cosmosante.NewRejectMessagesDecorator(
 			[]string{
 				sdk.MsgTypeURL(&evmtypes.MsgEthereumTx{}),
+				sdk.MsgTypeURL(&conntypes.MsgConnectionOpenInit{}), // don't let any connection open from the Rollapp side (it's still possible from the other side)
```
[the remark this answers is not in the document]
Comment (danwt) [reply app/ante/handlers.go:47]: I think ante is fine
If you think that's better I will think, anyways would prefer to change it in another PR

File: README.md [hunk 7]
```diff
@@ -222,25 +201,19 @@ After successful run, the new established channels will be shown
...
 
 ```shell
-sed -i 's/empty_blocks_max_time = "3s"/empty_blocks_max_time = "3600s"/' ${ROLLAPP_HOME_DIR}/config/dymint.toml
-```
-
-Mac:
-
-```shell
-sed -i '' 's/empty_blocks_max_time = "3s"/empty_blocks_max_time = "3600s"/' ${ROLLAPP_HOME_DIR}/config/dymint.toml
+dasel put -f "${ROLLAPP_HOME_DIR}"/config/dymint.toml "max_idle_time" -v "1h"
```
[the remark this answers is not in the document]
Comment (danwt) [reply README.md:210]: sure sure, out of scope

File: README.md [hunk 8]
```diff
@@ -169,32 +162,18 @@ sh scripts/update_genesis_file.sh
 Validate genesis file:
 
 ```shell
-rollapp-evm validate-genesis
+$EXECUTABLE validate-genesis
 ```
 
 ```shell
-# this script automatically adds 2 vesting accounts, adjust the timestampts to your liking or skip this step
+# this script automatically adds 2 vesting accounts, adjust the timestamps to your liking or skip this step
```
[the remark this answers is not in the document]
Comment (danwt) [reply README.md:169]: 🤷

File: app/app.go [hunk 9]
```diff
@@ -563,6 +545,16 @@ func NewRollapp(
 	transferStack = transfer.NewIBCModule(app.TransferKeeper)
 	transferStack = claims.NewIBCMiddleware(*app.ClaimsKeeper, transferStack)
 	transferStack = erc20.NewIBCMiddleware(app.Erc20Keeper, transferStack)
+	transferStack = hubgenkeeper.NewIBCModule(
+		transferStack,
+		func(ctx sdk.Context, transfer *ibctransfertypes.MsgTransfer) error {
```
Comment (zale144) [inline app/app.go:550]: can we pls use the original signature instead of wrapping

File: scripts/update_genesis_file.sh [hunk 10]
```diff
@@ -15,70 +15,16 @@ jq '.consensus_params["block"]["max_bytes"] = "3145728"' "$GENESIS_FILE" >"$tmp"
...
-jq --argjson elevated_address_json "$elevated_address_json" '.app_state.hubgenesis.params.genesis_triggerer_allowlist += $elevated_address_json' "$GENESIS_FILE" >"$tmp" && mv "$tmp" "$GENESIS_FILE"
+# ---------------------------- add genesis accounts for the hub ---------------------------- #
+genesis_accounts=$(cat "$ROLLAPP_SETTLEMENT_INIT_DIR_PATH"/genesis_accounts.json)
+jq --argjson genesis_accounts "$genesis_accounts" '.app_state.hubgenesis.state.genesis_accounts = $genesis_accounts' "$GENESIS_FILE" >"$tmp" && mv "$tmp" "$GENESIS_FILE"
 
 # ---------------------------- add denom metadata ---------------------------- #
 denom_metadata=$(cat "$ROLLAPP_SETTLEMENT_INIT_DIR_PATH"/denommetadata.json)
+elevated_address=$("$EXECUTABLE" keys show "$KEY_NAME_ROLLAPP" --keyring-backend test --output json | jq -r .address)
 jq --argjson denom_metadata "$denom_metadata" '.app_state.bank.denom_metadata = $denom_metadata' "$GENESIS_FILE" >"$tmp" && mv "$tmp" "$GENESIS_FILE"
 jq --arg elevated_address "$elevated_address" '.app_state.denommetadata.params.allowed_addresses += [$elevated_address]' "$GENESIS_FILE" >"$tmp" && mv "$tmp" "$GENESIS_FILE"
```
Comment (zale144) [inline scripts/update_genesis_file.sh:29]: I don't think we have `denom_metadata.params...` anymore

File: app/app.go (hunk 9 above)
Comment (danwt) [reply app/app.go:549]: sure I will change

File: scripts/update_genesis_file.sh (hunk 10 above)
Comment (danwt) [reply scripts/update_genesis_file.sh:29]: do we or don't we?

Comment (danwt) [reply scripts/update_genesis_file.sh:29]: I'll leave it to you to finish this please @zale144 
As it corresponds to your denom meta data work

**Review by mtsitrin — APPROVED** [review]

File: scripts/update_genesis_file.sh [hunk 11]
```diff
@@ -15,70 +15,16 @@ jq '.consensus_params["block"]["max_bytes"] = "3145728"' "$GENESIS_FILE" >"$tmp"
...
-elevated_address=$("$EXECUTABLE" keys show "$KEY_NAME_ROLLAPP" --keyring-backend test --output json | jq -r .address)
-elevated_address_json=$(jq -n \
-  --arg address "$elevated_address" \
-  '[{
-        "address": $address
-    }]')
-jq --argjson elevated_address_json "$elevated_address_json" '.app_state.hubgenesis.params.genesis_triggerer_allowlist += $elevated_address_json' "$GENESIS_FILE" >"$tmp" && mv "$tmp" "$GENESIS_FILE"
+# ---------------------------- add genesis accounts for the hub ---------------------------- #
+genesis_accounts=$(cat "$ROLLAPP_SETTLEMENT_INIT_DIR_PATH"/genesis_accounts.json)
+jq --argjson genesis_accounts "$genesis_accounts" '.app_state.hubgenesis.state.genesis_accounts = $genesis_accounts' "$GENESIS_FILE" >"$tmp" && mv "$tmp" "$GENESIS_FILE"
```
Comment (danwt) [inline scripts/update_genesis_file.sh:23]: @danwt

**Pull request merged**